### ISSUE TITLE
[BUG][xpu] TRITONXPU_COMPARE_FUSION=1 时 cmpf+masked store 在 tile≥4096 编译崩溃（triton_xpu.cmpf result 类型被判成 ui32）

### ISSUE THREAD

**Issue by JosephNew** [flagos-ai/FlagTree#1249]
### 环境

| 项 | 值 |
| --- | --- |
| FlagTree | `0.7.0+xpu.git803ea98b`（0.7.0-rc2 分支 HEAD，xpu backend，源码编译） |
| triton | `3.6.0`（FlagTree xpu fork） |
| torch | `2.9.0+cu128` |
| 硬件 | 昆仑芯 P800 OAM 96GB（宿主驱动 5.37.1，容器内 CUDA 兼容层 515.58） |
| 宿主 | `10.1.14.218`，容器 `klx-sglang-rc1` |

纯 Triton，不涉及 flag_gems / sglang。

### 现象

`TRITONXPU_COMPARE_FUSION=1` 时，一次 `arith.cmpf` 的结果喂给 masked store，**tile ≥ 4096 编译失败**：

```
OutOfResources: out of resource: uni_sram
PassManager::run failed: 'triton_xpu.cmpf' op result #0 must be floating-point or ranked tensor
of floating-point values ... but got
'tensor<16384xui32, #triton_xpu.cluster<{sizePerCore = [256], coresPerGroup = [64],
 groupsPerCluster = [1], order = [0]}>>'
 → 失败 pass：TritonXPUNormalize（dump-flag=false compare-fusion=true）
```

注意 `uni_sram` 那句是伴随症状：`Required: 0, Hardware limit: 0`，真正的报错是 `triton_xpu.cmpf` 的 result 类型被判成 `ui32`（应该是 bool/i1）。

**同一份 kernel、同一个进程，开关只差一个环境变量**：

| tile | `COMPARE_FUSION=1` | `COMPARE_FUSION=0` |
| --- | --- | --- |
| 256 | ✅ | ✅ |
| 512 | ✅ | ✅ |
| 1024 | ✅ | ✅ |
| 2048 | ✅ | ✅ |
| **4096** | ❌ `OutOfResources: uni_sram` | ✅ |
| **8192** | ❌ | ✅ |
| **16384** | ❌ | ✅ |
| **32768** | ❌ | ✅ |

阈值卡在 2048/4096 之间。

### 最小复现

```python
#!/usr/bin/env python3
"""Minimal repro: FlagTree xpu `compare-fusion` miscompiles compare+masked-store."""
import os

os.environ["TRITONXPU_COMPARE_FUSION"] = os.environ.get("COMPARE_FUSION", "1")
os.environ.setdefault("TRITON_CACHE_DIR", "/tmp/cf_repro")

import torch
import triton
import triton.language as tl


@triton.jit
def cmp_store_kernel(in0_ptr, in1_ptr, out0_ptr, num_tasks, TILE: tl.constexpr):
    pid = tl.program_id(0)
    tid = pid * TILE + tl.arange(0, TILE)
    mask = tid < num_tasks
    in0 = tl.load(in0_ptr + tid, mask=mask)
    in1 = tl.load(in1_ptr + tid, mask=mask)
    tl.store(out0_ptr + tid, in0.to(tl.float32) >= in1, mask=mask)


def run(tile):
    x = torch.randn(tile, device="cuda", dtype=torch.float32)
    y = torch.randn(tile, device="cuda", dtype=torch.float32)
    out = torch.empty(tile, device="cuda", dtype=torch.bool)
    try:
        cmp_store_kernel[(1,)](x, y, out, tile, TILE=tile, num_warps=16)
        torch.cuda.synchronize()
        print("[ OK  ] tile=%-6d correct=%s" % (tile, bool(torch.equal(out, x >= y))))
    except Exception as exc:
        first = str(exc).splitlines()[0] if str(exc) else ""
        print("[FAIL ] tile=%-6d %s: %s" % (tile, type(exc).__name__, first[:200]))


print("flagtree", triton.__version__,
      "TRITONXPU_COMPARE_FUSION=%r" % os.environ["TRITONXPU_COMPARE_FUSION"])
for tile in (512, 4096):
    run(tile)
```

运行（本机 2 次独立空缓存，结果一致）：

```bash
# 复现
TRITON_CACHE_DIR=/tmp/cf_repro python3 compare_fusion_repro.py
# 对照
COMPARE_FUSION=0 TRITON_CACHE_DIR=/tmp/cf_repro_off python3 compare_fusion_repro.py
```

实测输出：

```
# run 1 (TRITONXPU_COMPARE_FUSION='1')
flagtree 3.6.0 TRITONXPU_COMPARE_FUSION='1'
[ OK  ] tile=512    correct=True
[FAIL ] tile=4096   OutOfResources: out of resource: uni_sram PassManager::run failed: 'triton_xpu.cmpf' op result #0 must be floating-point ...

# 对照 (TRITONXPU_COMPARE_FUSION='0')
flagtree 3.6.0 TRITONXPU_COMPARE_FUSION='0'
[ OK  ] tile=512    correct=True
[ OK  ] tile=4096   correct=True
```

### 影响面（不是纸面问题）

FlagGems 昆仑芯 backend 的整套比较算子都在调用前后手动开关这个环境变量，`_kunlunxin/ops/` 下 9 个文件：

`eq.py` / `ge.py` / `greater.py` / `gt.py` / `le.py` / `less_equal.py` / `lt.py` / `ne.py` / `not_equal.py`

```python
# flag_gems/runtime/backend/_kunlunxin/ops/ge.py（节选）
def ge(A, B):
    logger.debug("GEMS_KUNLUNXIN GE")
    os.environ["TRITONXPU_COMPARE_FUSION"] = "1"
    os.environ["TRITONXPU_FP16_FAST"] = "1"
    res = ge_func(A, B)
    del os.environ["TRITONXPU_COMPARE_FUSION"]
    del os.environ["TRITONXPU_FP16_FAST"]
    return res
```
（这套开关来自 FlagGems 已合入的 #1032 `[KUNLUNXIN] Speed Up ge/gt/le/lt/ne with TRITONXPU_COMPARE_FUSION and TRITONXPU_FP16_FAST`；PR #6442 又在同一批算子上继续优化。）

实际后果：**sglang 0.5.18 在昆仑芯上的温度采样走不通**。temp=0.7 时比较算子吃到 vocab 量级的张量（≥ 4096），直接在这个 pass 上崩，起服务之后的第 2 个功能用例就挂，bench / MoE / VL 全部排不下去。SVT 侧 FlagOS 2.2 rc2 手工测试当前就卡在这一条。

临时绕法：把上面 9 组算子的注册名（`ge`/`ge_scalar`/`gt`/`gt_scalar`/`le`/`le_scalar`/`lt`/`lt_scalar`/`eq`/`eq_scalar`/`ne`/`ne_scalar`/`greater`/`greater_equal`/`less`/`less_equal`/`not_equal`）加进 `SGLANG_FL_FLAGOS_BLACKLIST`，退回 torch_xmlir 原生化——**实测有效**：27B tp=4 起服务后 temp=0.0/0.7/1.0 三个用例全 PASS，server 日志 `PassManager::run failed` 命中 0 次。（这 17 项一加，`ge`/`gt`/`le`/`lt` 在昆仑芯上的提速收益也就没了，属于要一起权衡的 trade-off。）

### 附：另一条独立缺陷

同一次排查里还发现 `tl.exp2` 在本 backend 被降级成 `exp`（静默错结果）+ block≥1024 编译崩，与本 issue 不是同一个 pass，另开：flagos-ai/FlagTree#1248。

**Comment by JosephNew** [comment]
### 复验：rc2 分支最新 HEAD 上仍然复现

把 FlagTree 升到 `0.7.0-rc2-triton3.6` 分支 HEAD `2372c4090`（`0.7.0+xpu.git2372c409`，2026-09-19，源码重编，pybind11 3.0.1）后重跑本 issue 的复现脚本，**现象完全一致**：

```
TRITONXPU_COMPARE_FUSION='1'
[ OK  ] tile=512    correct=True
[FAIL ] tile=4096   OutOfResources: out of resource: uni_sram ... 'triton_xpu.cmpf' op result #0 ...

COMPARE_FUSION=0（对照）
[ OK  ] tile=512    correct=True
[ OK  ] tile=4096   correct=True
```

本次升级区间（`803ea98b` → `2372c409`）只有 PR #1217 一个 commit（do_bench 除零守卫），与本缺陷无关，结果符合预期——列在这里是给复验留个基线记录。

复现脚本与命令见正文，未做任何修改。
